File: Tests.FSharp/CompoundPatternTests.fs
namespace Tests.FSharp

open System.Text.Patterns
open Microsoft.VisualStudio.TestTools.UnitTesting

[<TestClass>]
type CompoundPatternTests() =
    [<TestMethod>]
    member this.Identifier() =
        let pattern = Pattern.Letter >> +(Pattern.Letter || Pattern.DecimalDigitNumber || '_')
        
        let mutable result = pattern.Consume("hello")
        Assert.That.Captures("hello", result)
        
        result <- pattern.Consume("example_name")
        Assert.That.Captures("example_name", result)

        result <- pattern.Consume("_fail")
        Assert.That.Fails(result)

    [<TestMethod>]
    member this.IPv4Address() =
        let digit = (~~(('1' || '2') >> Pattern.DecimalDigitNumber) || Pattern.DecimalDigitNumber) >> ~~Pattern.DecimalDigitNumber
        let address = digit >> '.' >> digit >> '.' >> digit >> '.' >> digit

        
        Assert.That.Captures("1", digit.Consume("1"))
        Assert.That.Captures("11", digit.Consume("11"))
        Assert.That.Captures("111", digit.Consume("111"))

        let mutable result = address.Consume("192.168.1.1")
        Assert.That.Captures("192.168.1.1", result)

    [<TestMethod>]
    member this.NumberInBracket() =
        let number = '[' >> Pattern.Number >> ']'
        let numberList = +(number >> ~~(+Pattern.SpaceSeparator)) >> "[c]"
        Assert.That.Captures("[1] [2] [c]", numberList.Consume("[1] [2] [c]"))


    [<TestMethod>]
    member this.PhoneNumber() =
        let number = Pattern.Number * 3 >> '-' >> Pattern.Number * 3 >> '-' >> Pattern.Number * 4
        let result = number.Consume("555-555-5555")
        Assert.That.Succeeds(result)
        Assert.That.Captures("555-555-5555", result)

    [<TestMethod>]
    member this.WebAddress() =
        let protocol = "http" >> ~~'s' >> "://"
        let host = +(Pattern.Letter || Pattern.Number || '-') >> '.' >> ~~(+(Pattern.Letter || Pattern.Number || '-') >> '.') >> Pattern.Letter * 3
        let location = +('/' >> +(Pattern.Letter || Pattern.Number || '-' || '_'))
        let address = ~~protocol >> host >> ~~location
        let mutable result = address.Consume("www.google.com")
        Assert.That.Captures("www.google.com", result)
        result <- address.Consume("http://www.google.com")
        Assert.That.Captures("http://www.google.com", result)
        result <- address.Consume("http://www.google.com/about")
        Assert.That.Captures("http://www.google.com/about", result)

File: Tests.VB/CompoundPatternTests.vb
Imports System.Text.Patterns
Imports Microsoft.VisualStudio.TestTools.UnitTesting

<TestClass>
Public Class CompoundPatternTests
	<TestMethod>
	Sub Identifier()
		Dim Identifier = Pattern.Letter And +(Pattern.Letter Or Pattern.DecimalDigitNumber Or "_")
		Assert.That.Captures("hello", Identifier.Consume("hello"))
		Assert.That.Captures("example_name", Identifier.Consume("example_name"))
		Assert.That.Fails(Identifier.Consume("_fail"))
	End Sub

	<TestMethod>
	Sub IPv4Address()
		Dim Digit = (-((CType("1", Pattern) Or "2") And Pattern.DecimalDigitNumber) Or Pattern.DecimalDigitNumber) And -Pattern.DecimalDigitNumber
		Dim Address = Digit And "." And Digit And "." And Digit And "." And Digit
		Dim Result = Address.Consume("192.168.1.1")
		Assert.That.Captures("192.168.1.1", Result)
	End Sub

	<TestMethod>
	Sub PhoneNumber()
		Dim Number = Pattern.Number * 3 And "-" And Pattern.Number * 3 And "-" And Pattern.Number * 4
		Dim Result = Number.Consume("555-555-5555")
		Assert.That.Captures("555-555-5555", Result)
	End Sub

	<TestMethod>
	Sub WebAddress()
		Dim Protocol = "http" And -CType("s", Pattern) And "://"
		Dim Host = +(Pattern.Letter Or Pattern.Number Or "-") And "." And -(+(Pattern.Letter Or Pattern.Number Or "-") And ".") And Pattern.Letter * 3
		Dim Location = +("/" And +(Pattern.Letter Or Pattern.Number Or "-" Or "_"))
		Dim Address = -Protocol And Host And -Location
		Dim Result = Address.Consume("www.google.com")
		Assert.That.Captures("www.google.com", Result)
		Result = Address.Consume("http://www.google.com")
		Assert.That.Captures("http://www.google.com", Result)
		Result = Address.Consume("https://www.google.com/about")
		Assert.That.Captures("https://www.google.com/about", Result)
	End Sub
End Class
